### PR TITLE
stats/opencensus: OpenCensus instrumentation api

### PR DESCRIPTION
**Pull request by zasweq**
This PR shows the API for OpenCensus instrumentation in grpc-go. On the client side, an interceptor and stats handler will be part of the dial option, and on the server side, only a stats handler will be registered. The data passed in is the Sampler, to give users a knob on the trace sampling rate, and also a bool to disable traces selectively, needed for observability GA.

RELEASE NOTES: N/A

### CODE REVIEW

File: stats/opencensus/opencensus.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+ * limitations under the License.
+ */
+
+// Package opencensus implements opencensus instrumentation code for gRPC-Go
+// clients and servers.
+package opencensus
+
+import (
+	"context"
+	"go.opencensus.io/trace"
```
Comment (dfawley) [inline stats/opencensus/opencensus.go]: Nit: please organize your imports; something like:

```go
import (
	"std"
	"lib"
	"imports"

	"other"
	"imports"

	renamed "imports"
)
```

File: stats/opencensus/opencensus.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+//
+// Using this option will always lead to instrumentation, however in order to
+// use the data an exporter must be registered with the OpenCensus trace package
+// for traces and the OpenCensus view package for metrics. Client Side has
+// retries, so a Unary and Streaming Interceptor are registered to handle per
+// RPC traces/metrics, and a Stats Handler is registered to handle per RPC
+// attempt trace/metrics. These three components registered work together in
+// conjunction, and do not work standalone.
+func DialOption(to traceOptions) grpc.DialOption {
+	return joinDialOptions(grpc.WithChainUnaryInterceptor(unaryInterceptor), grpc.WithChainStreamInterceptor(streamInterceptor), grpc.WithStatsHandler(&clientStatsHandler{to: to}))
```
Comment (dfawley) [inline stats/opencensus/opencensus.go:58]: In practice the interceptor will need to be a method on an object or a lambda, right?

File: stats/opencensus/opencensus.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	Ts trace.Sampler
+	// DisableTrace determines whether traces are disabled for an OpenCensus
+	// Dial or Server option. will overwrite any global option setting.
+	DisableTrace bool
+}
+
+// DialOption returns a dial option which enables OpenCensus instrumentation
+// code for a grpc.ClientConn.
+//
+// Client Applications interested in instrumenting their grpc.ClientConn should
```
Comment (dfawley) [inline stats/opencensus/opencensus.go:46]: Nit: Applications->applications

File: stats/opencensus/opencensus.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+// pass the dial option returned from this function as the first dial option to
+// grpc.Dial().
+//
+// Using this option will always lead to instrumentation, however in order to
+// use the data an exporter must be registered with the OpenCensus trace package
+// for traces and the OpenCensus view package for metrics. Client Side has
```
Comment (dfawley) [inline stats/opencensus/opencensus.go:52]: Nit: "Client-side" or "Client side"

File: stats/opencensus/opencensus.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+// ServerOption returns a server option which enables OpenCensus instrumentation
+// code for a grpc.Server.
+//
+// Server applications interested in instrumenting their grpc.Server should
+// pass the server option returned from this function as the first argument to
+// grpc.NewServer().
+//
+// Using this option will always lead to instrumentation, however in order to
+// use the data an exporter must be registered with the OpenCensus trace package
+// for traces and the OpenCensus view package for metrics. Server Side does not
```
Comment (dfawley) [inline stats/opencensus/opencensus.go:70]: Server-side / Server side

File: stats/opencensus/opencensus.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+// have retries, so a registered Stats Handler is the only component you need
+// for RPC traces/metrics.
```
Comment (dfawley) [inline stats/opencensus/opencensus.go:72]: "... so a registered Stats Handler is the only option that is returned"?

Also it is probably important to call out that (at least for now) it's not possible to use this in conjunction with another stats handler dial/server option.

**Review by dfawley — APPROVED** [review]

File: stats/opencensus/opencensus.go (hunk 1 above)
Comment (zasweq) [reply stats/opencensus/opencensus.go]: I always see a space between std lib imports and others. However, I sometimes see non grpc imports separated from grpc imports, and renamed imports never separate, only thing I've seen after gRPC imports are something like proto imports. I put spaces in between std lib, other, and grpc.

File: stats/opencensus/opencensus.go (hunk 2 above)
Comment (zasweq) [reply stats/opencensus/opencensus.go:58]: Discussed offline, yes, especially for any traces that need interceptor level (and who knows, perhaps metrics will eventually get configuration knobs that affect interceptor level logic as well). Perhaps when the data is needed I'll capture the data persisted I need by making these interceptors objects on the client/server stats handlers

File: stats/opencensus/opencensus.go (hunk 3 above)
Comment (zasweq) [reply stats/opencensus/opencensus.go:46]: Whoops, switched.

File: stats/opencensus/opencensus.go (hunk 4 above)
Comment (zasweq) [reply stats/opencensus/opencensus.go:52]: Chose later. Changed Server Side later in file as well.

File: stats/opencensus/opencensus.go (hunk 6 above)
Comment (zasweq) [reply stats/opencensus/opencensus.go:72]: Fair enough, will add, but this was never possible right? I added binary logger option in this PR: https://github.com/grpc/grpc-go/pull/5675, which added a second possible binary logger on top of the global at each binary log recording point.

File: stats/opencensus/opencensus.go (hunk 1 above)
Comment (dfawley) [reply stats/opencensus/opencensus.go]: Yeah, there are no hard-set rules everyone follows, and without a linter for it, we'll never enforce any particular rule in our repo.  It's definitely a widely used convention to separate stdlib imports from others, so as long as we do that I'm fine with whatever other preferences people have.

Comment (zasweq) [reply stats/opencensus/opencensus.go]: Yeah that seems like the most important/always followed distinction.

File: stats/opencensus/opencensus.go (hunk 6 above)
Comment (dfawley) [reply stats/opencensus/opencensus.go]: Correct, it was never possible.  But this is the first time we've ever provided a function that returns a `grpc.WithStatsHandler` `DialOption` (/ `ServerOption`), as opposed to the OC package which provides a `grpc/stats.Handler`, so it's important to mention that this sets a `stats.Handler` and no other one should be used on the same client/server (for now).  It seems like we will need to support multiple in the not-too-distant future for migration to OpenTelemetry, though.

**Review by dfawley — APPROVED** [review]

**Pull request merged**